btree/btree.c

```diff
@@ -2684,6 +2696,7 @@ static void ff_init(const struct segaddr *addr, int ksize, int vsize, int nsize,
 	 * other place should the opaque data get captured and written to BE
 	 * segment.
 	 */
+	M0_BTREE_TX_CAPTURE(tx, seg, &h->ff_opaque, sizeof(h->ff_opaque));
```

Comment:
I am not sure if this will be the right place to capture opaque pointer, because of following two reasons:
1. We perform node initializaion even during the bnode_alloc(), but sometimes we might not use the allocated node by bnode_alloc() [in cases where we allocated nodes by seeing possibility of overflow and then don't use it]. So, in such cases we are capturing the region unnecessary.

2. In current implementation, we are calling bnode_capture() explicitly for root nodes dureing tree create, so capturing opaque pointer can be handled there.

Reply:
> This will only take care of the opaque pointer capture for root nodes but it does not take care of the capture for the rest of the nodes allocated for this tree.
For rest of the nodes, currently we are capturring the opaque pointer when we are actually using that node. For example, We allocate new node for spliting operation, allocate -> initialize -> insert new records -> call captured for capturing node header and rest of the things.
But for root node, user allocate -> btree does node initialization -> capture node header.

Reply:
However, I agree with your first point of capturing extra 8 bytes in some cases.

Reply:
> For rest of the nodes, currently we are capturring the opaque pointer when we are actually using that node. For example, We allocate new node for spliting operation, allocate -> initialize -> insert new records -> call captured for capturing node header and rest of the things. But for root node, user allocate -> btree does node initialization -> capture node header.

No we are not capturing opaque for the rest of the nodes. There was a bug in the code where the opaque pointer was never captured in the capture routines and that is because the opaque pointer would never be NULL. It would always contain pointer to the node descriptor by the time the capture routine is called. The opaque pointer happened to be NULL simply because the BE allocator sets all the bytes to NULL as a part of the allocation. With this new code where the BE allocator skips initializing the allocated memory we have non-NULL values in the opaque pointer and hence an explicit call to capture the opaque pointer is done.

Reply:
Got the answer